### PR TITLE
Allow processing of kinesis events to fallback top individual processing

### PR DESCRIPTION
**Pull request by jlogeart**
This change introduces the ability to reprocess events individually in case the batch processing failed.

It prevents events that are following a failing one to be discarded. 
A common use case is for instance processing a batch and shutting down before checkpointing. When consumption restarts, a new batch is taken from the queue, potentially with new events that have not yet been processed and the whole batch fails because of duplicate id, leading to the newly taken events to be discarded instead of processed.

Note this change is not source compatible with the previous version and will require careful deploy.

### CODE REVIEW

File: app/io/flow/event/v2/KinesisStyle.scala [hunk 1]
```diff
@@ -58,27 +60,76 @@ trait KinesisStyleRecordProcessor extends MetricsInstrumentation {
...
+              .warn(s"[FlowKinesisWarn] Error while processing records after $MaxRetries attempts. " +
+                s"$size records are retried individually. " +
+                s"Sequence numbers: ${records.map(_.sequenceNumber).mkString(", ")}", e)
+
+            retryIndividually(f, records)
+          } else {
+            logger
+              .withKeyValue("retries", retries)
+              .warn(s"[FlowKinesisError] Error while processing records after $MaxRetries attempts. " +
+                s"$size records are not retried individually are skipped. " +
```
Comment (benwaffle) [inline app/io/flow/event/v2/KinesisStyle.scala:81]:
```suggestion
                s"$size records are not retried individually, they are skipped. " +
```

File: app/io/flow/event/v2/KinesisStyle.scala [hunk 2]
```diff
@@ -58,27 +60,76 @@ trait KinesisStyleRecordProcessor extends MetricsInstrumentation {
...
+                s"Sequence numbers: ${records.map(_.sequenceNumber).mkString(", ")}", e)
```
Comment (benwaffle) [inline app/io/flow/event/v2/KinesisStyle.scala:82]: You should use `withKeyValue` to output the sequence numbers

File: test/io/flow/event/v2/IntegrationQueueSpec.scala [hunk 3]
```diff
@@ -274,22 +274,22 @@ class IntegrationQueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helper
 
       val asdf = publishTestObject(producer, TestObject(id = "asdf"))
 
-      var failed = false
-      var consumed = false
+      val failed = new AtomicBoolean(false)
+      val consumed = new AtomicBoolean(false)
```
Comment (benwaffle) [inline test/io/flow/event/v2/IntegrationQueueSpec.scala:278]: If you're looking for thread safety, I think you need to get() failed and set() consumed in a single atomic transaction

File: test/io/flow/event/v2/IntegrationQueueSpec.scala [hunk 4]
```diff
@@ -304,27 +304,122 @@ class IntegrationQueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helper
...
+          throw new Exception("pretend crash")
+        } else if (recs.head.eventId == e2) {
+          // e2 is reprocessed individually
+          tries2.incrementAndGet()
+          consumed2.set(true)
+        }
+      }
+
+      eventuallyInNSeconds(120) {
+        // 3 as a batch, 2 individually
```
Comment (benwaffle) [inline test/io/flow/event/v2/IntegrationQueueSpec.scala:417]: should this be 0 individually?

File: test/io/flow/event/v2/IntegrationQueueSpec.scala [hunk 5]
```diff
@@ -304,27 +304,122 @@ class IntegrationQueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helper
...
+      val triesIndividually1 = new AtomicLong(0)
+      val tries2 = new AtomicLong(0)
+      val consumed2 = new AtomicBoolean(false)
+
+      queue.consume[TestEvent] { recs =>
+        if (recs.head.eventId == e1) {
+          if (recs.size == 2) {
+            // make sure the other one is e2
+            recs(1).eventId mustBe e2
+            triesBatch1.incrementAndGet()
```
Comment (benwaffle) [inline test/io/flow/event/v2/IntegrationQueueSpec.scala:354]: this test is confusing to read. shouldn't you be incrementing `tries2`?

is there a cleaner way to write it? something in this style?
```
recs.map(_.eventId) match {
  case Seq(e1, e2) => tries.batch.incrementAndGet(); throw new Exception("batch failed")
  case Seq(e1) => tries.e1.individual.incrementAndGet()
  case Seq(e2) => tries.e2.individual.incrementAndGet()
}
```

File: test/io/flow/event/v2/IntegrationQueueSpec.scala [hunk 6]
```diff
@@ -304,27 +304,122 @@ class IntegrationQueueSpec extends PlaySpec with GuiceOneAppPerSuite with Helper
...
+        consumed2.get() must be (true)
+      }
+
+      queue.shutdownConsumers()
+    }
+  }
+
+  "fallback to individual processing can be disabled" in {
+    withConfig { config =>
+      config.set("development_workstation.lib.event.test.v0.test_event.json.enableIndividualProcessingFallback", "false")
```
Comment (benwaffle) [inline test/io/flow/event/v2/IntegrationQueueSpec.scala:381]: if the test is identical except for this line, can we move the test into a function?

**Review by benwaffle — CHANGES_REQUESTED** [review]

**Comment by benwaffle** [comment]
Other questions:
1. Do kinesis event consumers fail a lot? I thought they usually just insert into a DB queue
2. Where is the source incompatibility?

File: test/io/flow/event/v2/IntegrationQueueSpec.scala (hunk 3 above)
Comment (jlogeart) [reply test/io/flow/event/v2/IntegrationQueueSpec.scala:278]: I am mostly interested in the `volatile` portion embedded in the `Atomic` such that the memory is properly flused between threads upon assignments.

I could have used `@volatile` too.

**Comment by jlogeart** [comment]
You can search for `"num_skipped_records"` in our logs to find occurrences: https://service.us2.sumologic.com/ui/#/search/create?id=ZaIy2nQ6GVQqB9IIe05AfEU2scEDiYj7VbvbE1es

File: test/io/flow/event/v2/IntegrationQueueSpec.scala (hunk 4 above)
Comment (jlogeart) [reply test/io/flow/event/v2/IntegrationQueueSpec.scala:417]: No, it is retried twice (1 retry) before fully failing.

File: test/io/flow/event/v2/IntegrationQueueSpec.scala (hunk 5 above)
Comment (jlogeart) [reply test/io/flow/event/v2/IntegrationQueueSpec.scala:354]: Good suggestion

**Review by benwaffle — APPROVED** [review]

**Pull request merged**